### PR TITLE
Add/sftp hosting data layer

### PR DESCRIPTION
**Pull request by glendaviesnz**
#### Changes proposed in this Pull Request

* Adds sftp data layer to allow easier control of errors and also to prepare for  handling of multiple users in future feature updates

#### Testing instructions

See #37431

### CODE REVIEW

File: client/state/data-layer/wpcom/sites/hosting/sftp-user.js [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+			method: 'POST',
+			path: `/sites/${ action.siteId }/hosting/ssh-user/reset-password`,
+			apiNamespace: 'wpcom/v2',
+			body: {},
+		},
+		action
+	);
+};
+
+const receiveAtomicSFTPUserSuccess = ( action, sftpUser ) =>
```
Comment (mmtr) [inline client/state/data-layer/wpcom/sites/hosting/sftp-user.js]: Given the API response is rather simple, I think we can parse it here and drop the `fromApi` usage:

```js
const receiveAtomicSftpUserSuccess = ( 
	{ siteId, userId }, 
	{ username, password: undefined }
) => updateAtomicSftpUser( siteId, userId, { username, password } );
```

Note that we don't do the `! response || ! response.username` check here, but I think we can assume the API will return always a response and a username. If that's not the case, then the `onError` callback will be triggered.

File: client/state/hosting/reducer.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+ *      }
+ * }
+ *
+ * @param  {Object} state  Current state
+ * @param  {Object} action Action payload
+ * @return {Object}        Updated state
+ */
+const sites = ( state = {}, action ) => {
+	const { siteId, userId } = action;
+	const site = state[ siteId ] ? state[ siteId ] : {};
```
Comment (mmtr) [inline client/state/hosting/reducer.js]: Could we use `keyedReducer` here?

```js
import { HOSTING_SFTP_USER_UPDATE } from 'state/action-types';
import { keyedReducer } from 'state/utils';

export const atomicHosting = ( state, { type, { userId, sftpUser } } ) =>
	type === HOSTING_SFTP_USER_UPDATE 
		? { sftpUsers: ...state, [ userId ]: sftpUser }
		: state;

export default keyedReducer( 'siteId', atomicHosting );
```

File: client/state/selectors/get-atomic-hosting-sftp-user.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+/**
+ * Returns the sftp user details for given site and user
+ *
+ * @param  {Object}  state   Global state tree
+ * @param  {Number}  siteId The ID of the site we're querying
+ * @param  {Number}  userId  The ID of the user we're querying
+ * @return {Object} SFTP user details for selected user
+ */
+export function getAtomicHostingSFTPUser( state, siteId, userId ) {
+	return get( state, [ 'atomicHosting', 'sites', siteId, 'sftpUsers', userId ], null );
```
Comment (mmtr) [inline client/state/selectors/get-atomic-hosting-sftp-user.js]: If we use `keyedReducer` as suggested above, this should change to:
```suggestion
	return get( state, [ 'atomicHosting', siteId, 'sftpUsers', userId ], null );
```

File: client/state/data-layer/wpcom/sites/hosting/sftp-user.js [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+	if ( ! response || ! response.username ) {
+		throw new Error( 'Retrieving or updating the sftp user was unsuccessful', response );
+	}
+
+	return response.password
+		? { username: response.username, password: response.password }
+		: { username: response.username };
+};
+
+const requestAtomicSFTPUser = action => {
```
[the remark this answers is not in the document]
Comment (glendaviesnz) [reply client/state/data-layer/wpcom/sites/hosting/sftp-user.js]: Fixed

File: client/state/data-layer/wpcom/sites/hosting/sftp-user.js [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+import { http } from 'state/data-layer/wpcom-http/actions';
+import { dispatchRequest } from 'state/data-layer/wpcom-http/utils';
+import { registerHandlers } from 'state/data-layer/handler-registry';
+import {
+	HOSTING_SFTP_USER_REQUEST,
+	HOSTING_SFTP_USER_CREATE,
+	HOSTING_SFTP_PASSWORD_RESET,
+} from 'state/action-types';
+import { errorNotice } from 'state/notices/actions';
+import { translate } from 'i18n-calypso';
```
[the remark this answers is not in the document]
Comment (glendaviesnz) [reply client/state/data-layer/wpcom/sites/hosting/sftp-user.js]: Fixed.

File: client/state/data-layer/wpcom/sites/hosting/sftp-user.js [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+			body: {},
+		},
+		action
+	);
+};
+
+const receiveAtomicSFTPUserSuccess = ( action, sftpUser ) =>
+	updateAtomicSFTPUser( action.siteId, action.userId, sftpUser );
+
+const sFTPUserError = ( { siteId, userId } ) => dispatch => {
```
[the remark this answers is not in the document]
Comment (glendaviesnz) [reply client/state/data-layer/wpcom/sites/hosting/sftp-user.js]: Nice, thanks for this tip. Done.

File: client/state/hosting/reducer.js (hunk 2 above)
Comment (glendaviesnz) [reply client/state/hosting/reducer.js]: Another handy tip - I had to use a slightly different setup to get it working, but still much simpler than my original.

File: client/state/selectors/get-atomic-hosting-sftp-user.js (hunk 3 above)
Comment (glendaviesnz) [reply client/state/selectors/get-atomic-hosting-sftp-user.js]: Done.

File: client/state/data-layer/wpcom/sites/hosting/sftp-user.js (hunk 1 above)
Comment (glendaviesnz) [reply client/state/data-layer/wpcom/sites/hosting/sftp-user.js]: Yes, my approach was probably overkill  - have simplified as suggested.

File: client/state/data-layer/wpcom/sites/hosting/sftp-user.js [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+import { dispatchRequest } from 'state/data-layer/wpcom-http/utils';
+import { registerHandlers } from 'state/data-layer/handler-registry';
+import {
+	HOSTING_SFTP_USER_REQUEST,
+	HOSTING_SFTP_USER_CREATE,
+	HOSTING_SFTP_PASSWORD_RESET,
+} from 'state/action-types';
+import { errorNotice } from 'state/notices/actions';
+import { translate } from 'i18n-calypso';
+import { updateAtomicSFTPUser } from 'state/hosting/actions.js';
```
[the remark this answers is not in the document]
Comment (glendaviesnz) [reply client/state/data-layer/wpcom/sites/hosting/sftp-user.js]: Fixed.

**Comment by glendaviesnz** [comment]
> In order to test this PR, I think we could use the [console dispatcher](https://wpcalypso.wordpress.com/devdocs/client/state/console-dispatch/README.md) in the browser console:


Running dispatch from the console doesn't fire the data layer api call for me either for some reason. I have now rebased both this and #37431 so should be easy to test now by checking out this PR and then merging #37431 into it.

File: client/state/data-layer/wpcom/sites/hosting/sftp-user.js [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+			body: {},
+		},
+		action
+	);
+};
+
+const receiveAtomicSftpUserSuccess = ( action, { username, password } ) =>
+	updateAtomicSftpUser( action.siteId, action.userId, { username, password } );
+
+const sFTPUserError = ( { siteId, userId } ) => [
```
Comment (mmtr) [inline client/state/data-layer/wpcom/sites/hosting/sftp-user.js]: Minor: 
```suggestion
const sftpUserError = ( { siteId, userId } ) => [
```

File: client/state/data-layer/wpcom/sites/hosting/sftp-user.js [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+			method: 'POST',
+			path: `/sites/${ action.siteId }/hosting/ssh-user`,
+			apiNamespace: 'wpcom/v2',
+			body: {},
+		},
+		action
+	);
+};
+
+const resetAtomicSFTPPassword = action => {
```
Comment (mmtr) [inline client/state/data-layer/wpcom/sites/hosting/sftp-user.js]: Minor: 
```suggestion
const resetAtomicSftpPassword = action => {
```

File: client/state/hosting/actions.js [hunk 10]
```diff
@@ -1,10 +1,42 @@
...
+	sftpUser,
+} );
+
+export const createAtomicSftpUser = ( siteId, userId ) => ( {
+	type: HOSTING_SFTP_USER_CREATE,
+	siteId,
+	userId,
+} );
+
+export const resetAtomicSFTPPassword = ( siteId, userId ) => ( {
```
Comment (mmtr) [inline client/state/hosting/actions.js]: Minor:
```suggestion
export const resetAtomicSftpPassword = ( siteId, userId ) => ( {
```

File: client/state/data-layer/wpcom/sites/hosting/sftp-user.js (hunk 8 above)
Comment (glendaviesnz) [reply client/state/data-layer/wpcom/sites/hosting/sftp-user.js]: haha, my global search and replace let me down yesterday :-)

**Pull request merged**